packages/vite/src/node/build.ts

```diff
@@ -643,36 +643,45 @@ export function resolveLibFilename(
   return `${name}.${format}.${extension}`
 }
 
-function resolveBuildOutputs(
+export function resolveBuildOutputs(
   outputs: OutputOptions | OutputOptions[] | undefined,
   libOptions: LibraryOptions | false,
   logger: Logger
 ): OutputOptions | OutputOptions[] | undefined {
   if (libOptions) {
-    const formats = libOptions.formats || ['es', 'umd']
-    if (
-      (formats.includes('umd') || formats.includes('iife')) &&
-      !libOptions.name
-    ) {
-      throw new Error(
-        `Option "build.lib.name" is required when output formats ` +
-          `include "umd" or "iife".`
-      )
+    const libFormats = libOptions.formats || ['es', 'umd']
+
+    if (!Array.isArray(outputs)) {
+      const requiresName =
+        libFormats.includes('umd') || libFormats.includes('iife')
+
+      if (requiresName && !libOptions.name) {
+        throw new Error(
+          'Option "build.lib.name" is required when output formats include "umd" or "iife".'
+        )
+      }
+
+      return libFormats.map((format) => ({ ...outputs, format }))
     }
-    if (!outputs) {
-      return formats.map((format) => ({ format }))
-    } else if (!Array.isArray(outputs)) {
-      return formats.map((format) => ({ ...outputs, format }))
-    } else if (libOptions.formats) {
-      // user explicitly specifying own output array
+
+    // By this point, we know "outputs" is an Array.
+    if (libOptions.formats) {
       logger.warn(
         colors.yellow(
-          `"build.lib.formats" will be ignored because ` +
-            `"build.rollupOptions.output" is already an array format`
+          '"build.lib.formats" will be ignored because "build.rollupOptions.output" is already an array format.'
         )
       )
     }
+
+    outputs.forEach((output) => {
+      if (['umd', 'iife'].includes(output.format!) && !output.name) {
+        throw new Error(
+          'Entries in "build.rollupOptions.output" must specify "name" when output formats include "umd" or "iife".'
+        )
+      }
+    })
```

Comment:
Rollup does output the following error during build. But I think it's good to have this, because this will output the error before build.

```
vite v3.1.0-beta.0 building for production...
✓ 1 modules transformed.
dist/index.html                   0.76 KiB
dist/my-lib-custom-filename.mjs   0.59 KiB / gzip: 0.36 KiB
You must supply "output.name" for UMD bundles that have exports so that the exports are accessible in environments without a module loader.
```